File: app/src/main/java/com/twilio/video/app/data/api/VideoAppServiceDelegate.java
package com.twilio.video.app.data.api;

import android.content.SharedPreferences;
import com.twilio.video.app.BuildConfig;
import com.twilio.video.app.data.Preferences;
import com.twilio.video.app.data.api.model.RoomProperties;
import com.twilio.video.twilioapi.TwilioApiUtils;
import io.reactivex.Single;

public class VideoAppServiceDelegate implements TokenService {
    private static final String APP_FLAVOR_TWILIO = "twilio";
    private static final String APP_FLAVOR_INTERNAL = "internal";

    private final SharedPreferences sharedPreferences;
    private final VideoAppService videoAppServiceDev;
    private final VideoAppService videoAppServiceStage;
    private final VideoAppService videoAppServiceProd;

    public VideoAppServiceDelegate(
            final SharedPreferences sharedPreferences,
            final VideoAppService videoAppServiceDev,
            final VideoAppService videoAppServiceStage,
            final VideoAppService videoAppServiceProd) {
        this.sharedPreferences = sharedPreferences;
        this.videoAppServiceDev = videoAppServiceDev;
        this.videoAppServiceStage = videoAppServiceStage;
        this.videoAppServiceProd = videoAppServiceProd;
    }

    @Override
    public Single<String> getToken(String identity, RoomProperties roomProperties) {
        final String env =
                sharedPreferences.getString(
                        Preferences.ENVIRONMENT, Preferences.ENVIRONMENT_DEFAULT);

        String appEnv = resolveAppEnvironment(BuildConfig.FLAVOR);
        VideoAppService videoAppService = resolveVideoAppService(env);
        return videoAppService.getToken(
                identity,
                roomProperties.getName(),
                appEnv,
                roomProperties.getTopology().getString(),
                roomProperties.isRecordParticipantsOnConnect());
    }

    private String resolveAppEnvironment(String appFlavor) {
        // Video App Service only accepts internal and production for app environment
        switch (appFlavor) {
            case APP_FLAVOR_TWILIO:
                return "production";
            case APP_FLAVOR_INTERNAL:
                return "internal";
            default:
                return "production";
        }
    }

    private VideoAppService resolveVideoAppService(String env) {
        switch (env) {
            case TwilioApiUtils.DEV:
                return videoAppServiceDev;
            case TwilioApiUtils.STAGE:
                return videoAppServiceStage;
            case TwilioApiUtils.PROD:
                return videoAppServiceProd;
            default:
                return videoAppServiceProd;
        }
    }
}

File: app/src/main/java/com/twilio/video/app/data/api/VideoAppServiceDelegate.kt
/*
 * Copyright (C) 2019 Twilio, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.twilio.video.app.data.api

import android.content.SharedPreferences
import com.twilio.video.app.data.Preferences.ENVIRONMENT
import com.twilio.video.app.data.Preferences.ENVIRONMENT_DEFAULT
import com.twilio.video.app.data.Preferences.RECORD_PARTICIPANTS_ON_CONNECT
import com.twilio.video.app.data.Preferences.RECORD_PARTICIPANTS_ON_CONNECT_DEFAULT
import com.twilio.video.app.data.Preferences.TOPOLOGY
import com.twilio.video.app.data.Preferences.TOPOLOGY_DEFAULT
import timber.log.Timber

class VideoAppServiceDelegate(
    private val sharedPreferences: SharedPreferences,
    private val videoAppServiceDev: VideoAppService,
    private val videoAppServiceStage: VideoAppService,
    private val videoAppServiceProd: VideoAppService
) : TokenService {

    override suspend fun getToken(identity: String?, roomName: String?): String {
        val topology = sharedPreferences.getString(
                TOPOLOGY,
                TOPOLOGY_DEFAULT)
        val isRecordParticipantsOnConnect = sharedPreferences.getBoolean(
                RECORD_PARTICIPANTS_ON_CONNECT,
                RECORD_PARTICIPANTS_ON_CONNECT_DEFAULT)
        val env = sharedPreferences.getString(
                ENVIRONMENT, ENVIRONMENT_DEFAULT)

        val videoAppService = resolveVideoAppService(env!!)
        Timber.d("app service env = $videoAppService")
        return videoAppService.getToken(
                identity,
                roomName,
                "production",
                topology,
                isRecordParticipantsOnConnect)
    }

    private fun resolveVideoAppService(env: String): VideoAppService {
        return when (env) {
            TWILIO_API_DEV_ENV -> videoAppServiceDev
            TWILIO_API_STAGE_ENV -> videoAppServiceStage
            else -> videoAppServiceProd
        }
    }
}
